Guidelines: Submission Tests (#151)
codewars/docs: content/recipes/authoring/kata-snippets/full-tests.md

- **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement user solution as an alias or wrapper around reference solution used by tests. Reference solution should not be possible to be called by any other part of the kata other than the tests. In particular, it should not be a global function.  
Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.

## Fixed Tests

Fixed tests are tests with predetermined inputs, and do not change between test runs.

- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).

> - Tests should check not only valid inputs, but **invalid inputs and edge conditions should be tested too** (if problem allows them). `null` inputs, empty arrays, empty strings, etc. all should be tested with fixed tests, unless kata description explicitly says that such inputs won't be provided and don't have to be handled by a solution. Random tests are not a replacement for carefully designed assertions for edge cases.

Reviewer: ```suggestion
- **Tests should check not only valid inputs but also invalid inputs and edge cases** (if the problem allows them). `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
```

> - If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before corresponding scenario is tested with random tests.

Reviewer: ```suggestion
- If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.
```

## Random Tests

> Random tests are not common testing practice in "real life" coding, and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass.

Reviewer: ```suggestion
Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.
```

> - **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by a chance.

Reviewer: ```suggestion
- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.
```

> - **Randomly generated inputs should be different every time the test suite is run** (with some exceptions, see below). Be sure to seed your RNGs appropriately, and use your random functions correctly, so values used to test user solution are unpredictable. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.

Reviewer: ```suggestion
- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
- In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.
```

Splitting this into two separate points since "random" inputs as per the XKCD comic and "random" inputs due to forgetting to seed the RNGs are two different types of mistake.

Author: Actually I did not consider the XKCD scenario here, because I think I've never seen it in a kata. Do you think it's really necessary?

My goal here was to say "know how to use your RNGs and random APIs", without being too language specific. Notes on `srand` and `time` I wanted to put in separate howto, something like "Writing random tests for [some language]".

Additionally, I am afraid that if we put "truly" here, people might overkill the problem with crypto rngs or external entropy sources like `/dev/urandom/`.

Reviewer 2: shouldn't a specific thing like that be relevant only to snipets provided for that specific language, rather?

> - **Order of random tests should be non-deterministic.** Not only inputs should be randomized, but their order too. It should not be the case that, for example, 10 valid, random inputs is tested as the first step, and after that, tests generate and verify 10 invalid inputs. Order of tested scenarios should be unpredictable too, and one possible technique is to generate and collect a set of random inputs for all required scenarios, and shuffle them.

Reviewer: ```suggestion
- **The order of random tests should be non-deterministic** especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria). Not only should inputs be randomized, but their order as well. It should not be the case that, for example, 10 valid, random inputs are tested in the first step followed by 10 invalid ones. The order of tested scenarios should be unpredictable too, and one possible technique is to generate and collect a set of random inputs for all required scenarios and shuffle them.
```

> - Under some rare circumstances, it's allowed to use so called **randomized tests** instead of fully random ones. For some types of problems (for example, kata related to chess, or problems with small set of possible inputs) it's too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use predefined, hardcoded set of inputs (and possibly expected outputs too), and before tests are run, shuffle them. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for the result of the test.

Reviewer: ```suggestion
- Under some rare circumstances, it's allowed to use so called **randomized tests** instead of fully random ones. For some types of problems (for example, kata related to chess, or problems with small set of possible inputs) it's too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too), and shuffle them before the tests are run. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
```

- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.
- When you are going to reject solutions based on their performance, **make sure that what you measure is what you want**. It's possible that most of the available time will be consumed not by user solution, but, for example, by your input generators, or reference solution. Make sure that poor performance of your test suite does not affect the final decision whether the submitted solution is performant enough or not!

## Tests with Additional Restrictions

Some kata may impose additional restrictions on submitted solutions unrelated to its extensional behavior. For example, code golf kata usually require submitted solutions to be shorter than X characters, while others may require the task to be realized without a certain library or language feature. Every such requirement must be tested explicitly, but authors should be aware that it is difficult (or even infeasible) to do so _in a proper manner_:

> - If the user solution should be inspected as source code, it can be found in `/home/codewarrior/solution.txt`.

Reviewer 3: ```suggestion
- If the user solution should be inspected as source code, it can be found in `/workspace/solution.txt`.
```
`/home/codewarrior/solution.txt` has been deprecated for a while. The old language versions have `/home/codewarrior` symlinked to `/workspace` so `/worspace/solution.txt` works with any language and version.

- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function.
Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.

## Fixed Tests

Fixed tests are tests with predetermined inputs and outputs, and do not change between test runs.

- **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s). Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.

> - **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.

Reviewer 4: `null` usually isn't like an empty array or an empty string. For most problems `null` is a language-specific implementation detail rather than a value that makes sense, so there's no reasonable result to return, for example, if a `String` is `null`. I assume that if there isn't valid output for some input, there won't be such input. Whether it should or doesn't have to be specified explicitly in such cases, `null` tests often don't make sense.

Author: How do you suggest it should be handled then? Should `null` inputs be considered a design issue with a task?

Reviewer 4: Actually, this looks correct overall, but I guess the question here is what "explicitly" is. If it says "you are given a string", it already means that it can't be `null` because a "`string`" (as a data type) may be `null` in some language, but a "string" (as a normal word) can't.
Another thing is that this sentence seems to encourage testing `null` input, while it's not normal in most cases.

Author: ```suggestion
- **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. If description states that solution should handle some specific inputs, like for example empty arrays, empty strings, etc. then a corresponding fixed test case should be added.
```

Do you think it's any better?

Reviewer 4: There's no need to mention empty arrays and strings in the description if they aren't exceptions that should be handled in a different way.
* Find the sum of the elements => nothing special about an empty array here, it should be tested.
* Find the shortest string => if the opposite isn't clear from return types in all available languages, there's no possible unique answer for an empty collection, so it's not tested by default.

Author: How would you describe here requirement of "returm minimum number from array, or - 1 if empty", in context of this paragraph? Should it be simply not allowed as a crappy requirement?

Reviewer 4: It doesn't make sense, so it's a bad requirement in general. It may be acceptable in lower level languages like C if the input is unsigned and when alternative ways have been analyzed. In more complicated cases it might be acceptable in a kata where the main focus is on math and calculations and where coding style is secondary, but even then that's not ideal.

Reviewer 4: > invalid inputs and edge cases if the problem allows them

Invalid input is something that's not expected by default, edge cases are just a special subset of normal input, so I'm not sure about grouping them together like this...

Author: So you mean that since description mentions them, it's not invalid? Yeah, makes sense. 👍

Reviewer 4: Now it looks like "edge cases" are opposed to "valid inputs". But it's like: "you are given a number (signed integer), make sure it's non-negative and check if it's a perfect square; if the number is negative, throw an InvalidArgumentException". Non-negative numbers are valid input; negative numbers probably are invalid because the possibility of passing a negative number probably is an implementation detail, but since they can be passed, they may be called valid, it's kind of ambiguous here; 0 is an edge case (and maybe 1 as well), but it's definitely valid.

Author: Yeah you are right, and it seems that my English is too poor to word this paragraph properly :( Is there anything you'd propose? Remove the whole bullet point, reword it, or replace "valid" with something else, like "expected inputs", "normal/regular inputs", idk? I am all ears :)

Reviewer 4: I've edited it, but I still don't like how it sounds. Edge cases probably shouldn't be grouped with special cases in one sentence.

Author: To me, it sounds really cool. Thanks!

## Random Tests

Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.

> - **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.

Reviewer 4: Random tests primarily are there to prevent cheating. If there's a small number of edge cases or if there's a requirement to do something in case of, let's say, negative numbers, random tests don't add much to the fixed tests. Of course, if a solutions returns `None` only for `-42` instead of all negative numbers, it's not really correct, but checking for `-42` is as easy as checking for any negative number, so if for some reason adding negative numbers to the random tests is more difficult than just extending the range, it may not be worth the effort.

Author: You are right, I think I meant "all types of inputs" here, and not specifically edge cases. For example, when task deals with fractions, make sure that random tests generate various types of (allowed) fractions: proper fractions, reducible fractions, fractions with `num` > `den`, etc. They should not hope that randomness will provide all possible cases.

Maybe a suggestion that "one/few instances of an edge case could be shuffled into the input set if needed" cold be useful? Or not?

Reviewer 4: Yes, maybe something like this.

Author: ```suggestion
- **Each requirement should have a couple of corresponding random test cases.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate all possible scenarios by chance. If solutions are required to handle any edge cases, one or a few instances of each such case can be shuffled into the input set.
```

Sounds OK?

Reviewer 4: Edge cases can be single values, like there's no point to test for `0!` in the random tests if it's already in the fixed tests.

Author: It generally should be in fixed tests, so maybe I should just remove this remark? Idk

Reviewer 4: For an average kata, complexity of random tests should be proportional to complexity of bypassing fixed tests and to potential reward obtained by those who do so (the rank of the kata). As random tests should normally be present when it's not possible to cover the whole input range by fixed tests, the random tests should try generating special subsets of input values if they are not fully covered by the fixed tests. Probably something like that.

Author: ```suggestion
- Random tests can be limited only to types of inputs, which are not covered with sample tests in the whole range. They should generate test cases for all scenarios which cannot be completely tested with fixed tests.
```

Maybe like this?

I am not sure how to go with a note on difficulty of working around them, because I am afraid that authors will be like "it's a 7 kyu kata, so it does not have to have good tests". What do you think?

Reviewer 4: This looks OK, but I suppose "sample tests" here is a mistake?

Author: `/s/sample/fixed/`. thanks for spotting.

- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
- **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.

> - **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.

Reviewer 4: Unless a kata relies mainly on the random tests by design, random tests shouldn't fail if no fixed tests failed, so debugging with random tests shouldn't be a thing in most cases, thus their output isn't too important.

Author: ```suggestion
- If it's not possible to build fixed tests in a way they'd be able to catch most of the failures and you have to rely on random tests to catch errors, **make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

Reviewer 2: @error256 yes but in some cases, it's pretty hard to actually figure out possible edge cases, because they will depend on the implemetation of the user. Alphametic solver, for example. You can imagine a lot of special cases up front, but considering the length of the code, it's impossible to ensure that you're covering everything. So a first batch of small random tests is still of interest (but not an excuse to not put as much egde cases as possible in the fixed tests).

Reviewer 4: Yes, but it can be considered a kata that relies on random testing.
Anyway, what's the preferred workflow here? Should individual suggestions be committed or it's better to batch-commit there in a group with other changes?

Author: For typos, we usually commit suggestions with fixes in batch. For paragraphs, we can commit single ones after we think they reached good state. If you see any suggestion addressing your remark and you think it's good, you can commit it without waiting for others.

You can also write your own suggestions or edit other suggestions to fix typos or wording.

  - Reference solution being accessible to users by mistake
  - Input mutation by the user solution affecting the input passed to the reference solution
  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function.
Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.

## Fixed Tests

Fixed tests are tests with predetermined inputs and outputs, and do not change between test runs.

> - **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).

Author: - **Each requirement should have a corresponding fixed test.** Every aspect of the specification in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s). Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.

## Random Tests

Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.

- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.
- **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.

> - In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.

Author: I think this part is too detailed and it rather belongs to some C howto. I added a note on proper use of random APIs above.

```suggestion
```

next: /recipes/authoring/kata-snippets/preloaded/
---

> # Writing Submit Tests

Reviewer 3: ```suggestion
# Writing Submission Tests
```

- **Make assertion messages clear.** Some testing frameworks are notorious for creating extremely unhelpful assertion messages. `Test failed: True should equal False`  is one example of such a message, and when the user encounters it, it's difficult for them to tell why the test failed. Check what messages are produced by your tests (especially failed ones) and make sure they are clear for users attempting the kata, and use assertions which allow for custom messages. Failed test cases should ideally present the input along with expected and actual values to the user. For example, for a prime testing kata, it could be something like `27: Expected False, but was True` which is much more informative.
- **Tests should handle unexpected answers gracefully.** Tests should not crash, and should present meaningful error messages when an unexpected answer is returned by the user's solution, e.g. `null` references when none is expected or an object of the incorrect type.

> - **Tests should not print excessively to standard output.** Console output is limited to 1.5 MB and when the limit is exceeded, the test suite is marked as failed and aborted by the runner. Users sometimes debug their solutions by printing to stdout, and when the test suite itself uses up a significant fraction of the limit, users may have not enough left for debugging. This applies especially to kata which output HTML to create some visualizations: such HTML uses up the buffer space allowed for stdout output. Ideally, the test suite should give detailed output **only when a test fails**. Successful tests should be as quiet as possible.

Author: This sentence was introduced by Donald, and I trust him more than us, that's why I did not accept your change :)

> - **Tests should not print excessively to standard output.** Console output is limited to 1.5 MB and when the limit is exceeded, the test suite is marked as failed and aborted by the runner. Users sometimes debug their solutions by printing to stdout, and when the test suite itself uses up a significant fraction of the limit, users may have not enough left for debugging. This applies especially to kata which output HTML to create some visualizations: such HTML uses up the buffer space allowed for stdout output. Ideally, the test suite should give detailed output **only when a test fails**. Successful tests should be as quiet as possible.

Reviewer 2: On this one, I disagree. ;p It's about meaning, not formulation.

## Reference Solution

Some test suites require a reference solution to generate the expected value(s) and compare them to those returned by the user solution. While such practice is common and sometimes necessary, there are things which have to be handled carefully:

- **Avoid using a reference solution if at all possible.** Sometimes there's no other way, but quite often a reference solution is simply not necessary. Most of the time, test cases can be generated with the answer known upfront which eliminates an entire class of potential problems associated with reference solutions:
  - Extra time spent on running a reference solution
  - Reference solution being accessible to users by mistake
  - Input mutation by the user solution affecting the input passed to the reference solution
  - Incorrect implementation of the reference solution leading to rejection of valid user solutions

> - **Reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.

Reviewer: ```suggestion
- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.
```

- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global function.
Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.

## Fixed Tests

Fixed tests are tests with predetermined inputs, and do not change between test runs.

- **Each requirement should have a corresponding fixed test.** Every requirement in the kata description should be explicitly tested with at least one, possibly more, dedicated and properly labeled assertion(s).

> - **Tests should check not only valid inputs but also invalid inputs and edge cases** (if the problem allows them). `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.

Author: ... but there's this _"if the problem allows them"_? This paragraph actually calls for input validation to be tested only if it's necessary, doesn't it?

> - **Tests should check not only valid inputs but also invalid inputs and edge cases** (if the problem allows them). `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.

Reviewer 2: ```suggestion
- **Tests should check not only valid inputs but also invalid inputs and edge cases if the problem allows them**. `null` inputs, empty arrays, empty strings, etc. should all be tested with fixed tests, unless the kata description _explicitly_ states that such inputs need not be considered. Random tests are not a replacement for carefully designed assertions for edge cases.
```

ok, so that's a readability issue. Suggesting this, then.

> - If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. Fixed tests are usually easier to debug, so ideally a fixed test should fail before the corresponding scenario is tested with random tests.

Reviewer 2: "make sure" -> it's actually not imperative, but strongly recommended. Sometimes it's hard to achieve...

=> soften the sentence?

Author: Maybe I could just remove it? It feels somewhat reduntand with the first bullet point, so they could be combined, I think...

```suggestion
```

## Random Tests

Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.

- **Each requirement should have at least one (preferably more) corresponding random test case.** Basically, randomly generated scenarios should correspond to the set of fixed tests. You should not rely on luck and hope that random tests will generate invalid inputs and edge conditions by chance.

> - **Random inputs should be _truly_ (pseudo-)random**, i.e. they should be _dynamically_ generated on every run with the help of suitable library functions such as `Math.random()` in JavaScript or `rand()` in C. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.

Reviewer 2: on this and the following: the XKCD has to be kept, but the C/C++ note might be too precise for this page (it's language specirfic). Not sure...

Author: Yup, I will think about it and get rid of this C thing.

Author: - **Use appropriate random utilities available in your language**. Know how to use RNGs, how to randomly generate types of inputs you need, be it numbers in a range, large numbers, strings, big integers, floating point values, collections.  Know how to do random sampling of collections, how to shuffle them, how to avoid duplicates. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.

- In some languages such as C/C++, **make sure that the RNGs are properly seeded with the current time at the beginning** using `srand(time(NULL))` or equivalent; otherwise, the same pseudorandom sequence will be generated on every run and your "random" inputs will devolve into fixed ones.

> - **The order of random tests should be non-deterministic** especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria). Not only should inputs be randomized, but their order as well. It should not be the case that, for example, 10 valid, random inputs are tested in the first step followed by 10 invalid ones. The order of tested scenarios should be unpredictable too, and one possible technique is to generate and collect a set of random inputs for all required scenarios and shuffle them.

Reviewer 2: this part needs a small rework:

"the order of the random tests" here is actually about "the order of the __outputs__". Hence saying clearly that "the kinds of outputs shouldn't be grouped together when random testing" seems far more... clear.

Author: Right, I just had no good idea how to word it. I keep this in mind.

Reviewer 2: this way?

```suggestion
- **Random tests shouldn't be grouped by kind of outputs or behaviours**, especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria), or when it comes to error checking (solution throwing an exception in some scpecific situations). Hence, the order of tested scenarios should be unpredictable too. One possible way to achieve this is to generate and collect a set of random inputs for all required scenarios and shuffle them before the actual tests.
```

note: `scenarii`?

Reviewer: > note: `scenarii`?

Nah, [Wikitionary](https://en.wiktionary.org/wiki/scenarii#English) says that such usage is rare in modern English.

> - Under some rare circumstances, it's allowed to use so called **randomized tests** instead of fully random ones. For some types of problems (for example, kata related to chess, or problems with small set of possible inputs) it's too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too), and shuffle them before the tests are run. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.

Reviewer 2: imo, the definitions were a good addition. Here, randomized tests aren't explained.

Mmmmmh, wait... iirc I put something in the glossary already

```suggestion
- Under some rare circumstances, it's allowed to use so called [**randomized tests**](/glossary/#randomized-tests) instead of fully random ones. For some types of problems (for example, kata simulating chess, or problems with small set of possible inputs) it may be too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too) and shuffle them before the tests are run. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test. Or a small fraction of the hardcoded inputs could be randomly skiped at each run to make the hardcoding of the results more tedious for the user.
```

- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.

> - **Difficulty should be consistent between test runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.

Reviewer 2: ```suggestion
- **Difficulty should be consistent between tests runs.** When randomized ranges are very large, it becomes possible for some users to receive many small, easy inputs while other users receive the exact opposite. Make sure that your random tests are built in a way which minimizes the chances for such a situation to occur. If you want to test difficult inputs, split the test cases into a set of easy ones and a set of difficult ones, and test them separately.
```

> - **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.

Reviewer 2: ```suggestion
- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because the input is very large and cannot be easily read. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
```

## Performance Tests

> Some kata require solutions to be fast, or to fall into some performance category. For example, the author may only wish to accept solutions completing in (sub-)linear time. Building such test suites is not an easy task!

Reviewer 2: ```suggestion
Some kata require solutions to be fast enough. For example, the author may only wish to accept solutions completing in (sub-)linear time. Building such test suites is not an easy task!
```

simplified, because too much redundant with the next sentence

> - **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.

Reviewer 2: ```suggestion
- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth to have a separate set of random tests with small inputs.
```

- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with small inputs.** For example, 200 tests with huge numbers or arrays is usually better than 10,000 tests with small ones (but see remarks on size of inputs above).
- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!
- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.

> - When you are going to reject solutions based on their performance, **make sure that what you measure is what you want**. It's possible that most of the available time will be consumed not by user solution, but, for example, by your input generators, or reference solution. Make sure that poor performance of your test suite does not affect the final decision whether the submitted solution is performant enough or not!

Reviewer 2: I still find that part not precise enough. So adding more details again (but less than before x) ). And suggesting again the thing about the slow solution (but in a softer way), because it's a general consideration, so it has imo its place in this "general documentation about testing performances". ;)

```suggestion
- **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performances. With huge inputs, the random generation may be more consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performances tests are actually discriminating the different kind of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generatin time.
- When it comes to maintainning a kata with performances requirements, it can be especially useful to have a solution whose the time complexity is supposed to be rejected on hand. Storing it, properly commented, in the test suite is then a very useful feature.
```

Author: ```suggestion
- **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performance. With huge inputs, the random generation may be more time-consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performance tests are actually discriminating the different kinds of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generation time.
```

> - **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!

Reviewer 2: ```suggestion
- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. Achieving this generally calls for very large inputs, so be careful!
```

"this" may relate in the mind of some users to the micro-optimizations, at first glance. Not sure it changes much with my suggestion, but avoiding "this" could be enough.

## Tests with Additional Restrictions

> Some kata may impose additional restrictions on submitted solutions unrelated to its extensional behavior. For example, code golf kata usually require submitted solutions to be shorter than X characters, while others may require the task to be realized without a certain library or language feature. Every such requirement must be tested explicitly, but authors should be aware that it is difficult (or even infeasible) to do so _in a proper manner_:

Reviewer 2: ```suggestion
Some kata may impose additional restrictions on the submitted solutions unrelated to its extensional behavior. For example, code golf kata usually require submitted solutions to be shorter than X characters, while others may require the task to be realized without a certain library or language feature. Every such requirement must be tested explicitly, but authors should be aware that it is difficult (or even infeasible) to do so _in a proper manner_:
```

> - **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performance. With huge inputs, the random generation may be more time-consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performance tests are actually discriminating the different kinds of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generation time.

Reviewer 2: ```suggestion
- **Make sure that what you measure is what you want**, when solutions are to be rejected based on their performance. With huge inputs, the random generation of the inputs may be more time-consuming than the computation of the expected result. Hence the overall timing indication in the output panel is generally useless to ensure that the performance tests are actually discriminating the different kinds of solutions as expected. The evaluation of the time actually used by the user's solution should be done (and compared to the reference solution if any) excluding the input generation time.
```

> - When maintaining a kata with performance requirements, it can be useful to have an access to a solution whose time complexity is supposed to be rejected. It can be used by maintainers to gauge the size of inputs for performance tests, to make sure they consistently fail. Storing it, properly commented, in the test suite, can be very helpful.

Reviewer 2: ```suggestion
- When maintaining a kata with performance requirements, it can be useful to have access to a solution whose time complexity is supposed to be rejected. It can be used by maintainers to gauge the size of inputs for performance tests, to make sure they consistently fail. Storing it, properly commented, in the test suite, can be very helpful.
```

# Writing Full Test Suite

This article contains a set of guidelines, which can be used by kata authors to create good test suites for their kata. They were collected to help to ensure that kata, translations, and their test suites are of sufficient quality and users' experience with tests will be as good as possible. Failing to comply with below guidelines should be reported as an issue with request for fix, preferrably during a review before the content gets approved. In case of severe violations, affected kata or translation can be retired, moved back to beta state, or rejected.

NOTE: There are many kinds of kata, and some guidelines might simply not apply to some of them. There are puzzles, hackmes, kata which mess with internals of language runtime, and some types of challenges require special ways of testing. But still, guidelines found below apply to majority of kata in some extent, and when used, kata author can be sure that their tests will be clear, efficient, reliable, and easy to debug.  

## General Guidelines

- Tests are code, and as such, they should **conform to [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/)**.

> - **Be familiar with the testing framework used by Codewars for your language.** Some languages use traditional, widely known testing frameworks, like `JUnit` for Java or `chai` for JavaScript. Some use frameworks dedicated specifically for use by Codewars (for example Python). Know what assertions are there, how to test objects, types, collections, etc. Know techniques provided by the framework (test case generators, parametrized test cases, etc.) and assertion libraries. You can visit the [reference page for your language](/languages/) to find more detailed information.

Author: I am not sure I get you right, which part exactly you think is not needed?

> - **Test suite should be organized** by utilizing API of your testing framework. Frameworks use `Describe` blocks, test fixtures, categories, or other means, to collect test cases into manageable blocks, label them, etc.

Author: Yes, I am trying to stay out of specifics (it's `Describe` for C test suites ;) ) because approaches differ between test frameworks.

C, Python, JS use granularity levels of describe/it/assert, for Java and C# it's test fixture/test method, for other languages it can be labelled categories or something else, and the levels do not always correspond between approaches. I want to stay away from this in my description and all I want to tell is "organize your tests". Specific hints should be possible to find in docs related to specific language or test framework reference.

> - **Test suite should be organized** by utilizing API of your testing framework. Frameworks use `Describe` blocks, test fixtures, categories, or other means, to collect test cases into manageable blocks, label them, etc.

Reviewer 2: ok :+1: (I was too much python/JS oriented, there)

- Usually the full test suite should contain **[fixed tests](#fixed-tests)**, and **[random tests](#random-tests)**. Some test frameworks offer features like test case generators (`QuickCheck`, `ScalaCheck`), and when used, authors can choose different structure for their test suite, but still they should create test specifications in such a way it covers all necessary scenarios. See specific sections below for more details.
- **Tests should not be susceptible to input mutation.** It should not be possible for user solution to modify test input in a way which would influence execution and result of the test. If reference solution is used, it should be called first, and user solution should be called afterwards with the same input.
- **Reference solution should not be revealed to user.** When assertion fails or test crashes, some test frameworks print fragment of source code which caused failure to the console. It may happen that such printed failure message or stack trace exposes information about solution which should not be revealed.

> - **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement user solution as an alias or wrapper around reference solution used by tests. Reference solution should not be possible to be called by any other part of the kata other than the tests. In particular, it should not be a global function.

Reviewer 2: ```suggestion
- **Reference solution shouldn't be accessible to user solution.** It should not be possible to implement user solution as an alias or wrapper around reference solution used by tests. Reference solution should not be possible to be called by any other part of the kata other than the tests. In particular, it should not be a global function (this problem may appear only in some languages, so look after the documentation of the language you're using).
```

> - **Test suite should not take long time to run.** For kata which are not performance oriented, size of the test suite should allow it to be completed in just a few (ideally 3-4) seconds. There's a time limit for every run (for some language  it is 12 seconds, for others it's 16 seconds), but generation of test cases, invovations of user solution and calls to reference solution (when used) should run well below this limit. Otherwise there's a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected.

Reviewer 2: ```suggestion
- **Test suite should not take long time to run.** For kata which are not performance oriented, size of the test suite should allow it to be completed in just a few seconds (ideally 2-3) seconds. There's a time limit for every run (for some language  it is 12 seconds, for others it's 16 seconds), but generation of test cases, invocations of user's solution and calls to reference solution (when used) should run well below this limit. Otherwise there's a risk that tests will inconsistently time out, or some valid, but not optimally performant solutions will be rejected. For performances oriented kata, be very careful that the generation of the inputs isn't the main time consuming part of the tests, or adapt specifically the way you build the random tests.
```

Reviewer 2: same here, it's a bit redundant with the lower part. But again, I think it could be good to have the info doubled...

## Fixed Tests

Fixed tests are tests with predetermined inputs, and do not change between test runs.

- **Each requirement should have corresponding fixed test.** Every requirement of the kata described in the description should be explicitly tested by at least one (possibly more) dedicated, properly labeled test case(s).

> - If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. When random test fails, it should be accompanied by at least one failed equivalent fixed test.

Reviewer 2: ```suggestion
- **Do not count on the random tests to catch edge cases:** If your test suite uses random tests (and it usually should, see below), then make sure that every test scenario generated there has its equivalent fixed test. When random test fails, it should be accompanied by at least one failed equivalent fixed test.
- **Boundaries should be tested extensively**, especially on kata where the output differs depending on a couple of conditions : exclusivness, inclusivness, behaviour above and lower the boundary need to be tested.
```

## Random Tests

> Random tests are not common testing practice and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds. Goal of random tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass. Random tests are necessary especialy for difficult, highly rewarding kata, but Codewars testing guidelines require them for every kata they are applicable for.

Reviewer 2: ```suggestion
Random tests are not common testing practice in industry and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hardcoding, and other workarounds.
The goal of  therandom tests is to make input of test cases unpredictable, so only solutions which actually solve the problem can pass. Random tests are necessary especialy for difficult, highly rewarding kata, and Codewars testing guidelines require them for every kata they are applicable for.
The lack of random tests is a straight condition for rejection of a beta kata or a translation.
```

Reviewer 2: I'm still not satisfied by the sentence:
>  Random tests are necessary especialy for difficult, highly rewarding kata, and Codewars testing guidelines require them for every kata they are applicable for.

It results in a weird impression, because on one side it's repuired, but then way pointing explicitely the hard kata?
Imo, this should be reworded one way or another

> - **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests.

Reviewer 2: ```suggestion

### Random tests: definition

Tests are considered random when they match the following behaviours:
* When the test suite is run one time, the inputs of the test are unpredictable.
* When the test suite is run one time, the outputs of the test are equally unpredictable.
* When the tests are run several times, the order of the inputs/outputs is different, and inputs/outputs showing up in one run may not appear in other runs.

Hence, different batches of random tests where one expects one kind of behavior (test only primes, or successful tests), then another expect another behaiour (test only composite numbers or expecting errors) aren't random tests.

### Randomized tests: definition

In some rare situations, it may be  either not possible or too complex to actually build a satistfaying random generator. Or even, the domain of the testable inputs is too small. In those specific **but rare** cases, randomized tests are acceptable. Tests are considered randomized when:
* The inputs (possibly the output) are hardcoded
* From on run to the other, only the order of the inputs/outputs vary

Keep in mind that randomized tests can easily be hardcoded so are to be avoided as much as possible, to the benefit of real random tests. One way to hide a bit the randomized testing could be to have a set of `N` inputs, and build the randomized tests by picking randomly only a fraction of those inputs. But again, true random tests are better.

### Guidelines for random testing

- **Each requirement should have at least one (preferrably more) corresponding random test case generated.** Basically, randomly generated scenarios should correspond to the set of fixed tests.
```

> - **Order of random tests should be non-deterministic.** Not only inputs should be randomized, but their order too. One possible technique is to generate and collect a set of random inputs for all required scenarios, and shuffle them.

Reviewer 2: ```suggestion
```

removed because of my big addition above

Reviewer 2: note, in the sentence just above; typo at `preferrably` -> `preferably`

Author: I did not go with two big, introductory definitions, but I added a point about "randomized tests". that was a good point I missed.
I think this paragraph is good to stay.

> - If possible, **reference solution should not be used**. Sometimes there's no other way, but quite often reference solution is simply not necessary. Test cases can be generated with the answer known upfront, and it can save time which would be spent on running the reference solution.

Reviewer 2: ```suggestion
- **The use of a reference solution should be avoided if possible**. Sometimes there's no other way, but quite often reference solution is simply not necessary. Test cases can be generated with the answer known upfront, and it can save time which would be spent on running the reference solution. This way of building the tests avoid a lot of problems (input mutation, security of the reference solution, ...).
```

we could put here again (in aother point) the part about the fact that the ref solution that mustn't be accessible to the user. I have in mind the fact that critical stuff shouldn't be put in preloaded.
That's redundant, I know. But better to have the info several times, considering htis kind of article is sort of the core of codewars (at least of what makes our day to day life as beta testers).

> - **Assertions should be close to the calls of tested solution**, preferrably immediatelly following it. Debugging becomes difficult if invalid answer is verified long after it's returned.

Reviewer 2: ```suggestion
- **Assertions should be close to the calls of tested solution**, preferrably immediatelly following it. Debugging becomes difficult if invalid answer is verified long after it has been returned.
```

- **Random tests should be run after fixed tests.** Not all testing frameworks allow for easy ordering of tests, but fixed tests should run, and eventually fail, before random tests.
- **Difficulty should be consistent beetwen test runs.** When randomized ranges are very large, then it's possible that one user receives many small, easy inputs and their solution passes, while another user receives many large, more difficult inputs, and their solution fails or times out. Make sure that your random tests are built in a way which elimnates such situation. If you want to test difficult inputs, split test cases into a set of easy ones and a set of diffficult ones, and tets them separately.

> - **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If possible, Try to compose your test suite in such a way that when test fail, they fail on some small, easy, debuggable input, before they fail on a large one, which is difficult to print. 

Reviewer 2: ```suggestion
- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. In such a case, two batches of random tests should be provided: a first batch with small debuggable inputs (note that the inputs are small, not the batch. So keep a reasonnable number of tests: around 50 at least), then the batch on wide inputs. The kind of inputs generated in both batches should match, so that the user can actually find the critical cases for his code in the small inputs.
```

## Performance Tests

Some kata require solutions to be fast, or to fall into some performance category. For example, author might want to accept only solutions of complexity `O(n)`, and reject `O(n^2)` ones. Building such test suite is not always easy!

> - **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.

Reviewer 2: ```suggestion
- **Performance tests must be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
```

Author: Why "must"? It's not that random tests and performance tests always have to be one and the same block? Block of random tests **can** be used to test performance, but why would it always have to be?

Reviewer 2: right :+1:

- If difficulty of a tested kata is related to the size of the input, it's usually **better to have a few tests with large input, that thousands of tests with small input.** 200 tests with very large numbers or arrays is usually better than 10_000 tests with small ones (but see remarks on size of inputs above).
- **Difference beetwen accepted and rejected solutions should be easy to spot.** Ideally, tests for accepted solutions should complete very fast, and tests for rejected solutions should take very long. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful! 
- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.

> 

Reviewer 2: ```suggestion
- **Be extra careful about the parformance of what part of code you actually test.** For some problems, the generation of the input is taking much more time than the actual comutation of the outputs, or takes at least a large amount of time. When performances tests are designed, the overall time consumed wy the tests as well as the timings possibly displays on the different blocks of the test suite **cannot be trusted** to evaluate how your tests are demanding **as long as they contain the inputs generation time**. Hence, inputs often must be computed first, outside of the test block, then the tests are run and you can finally compare two different codes.
- **Always store a version of the code with the time complexity to reject in the test suite.** this way, adjustments can be done easily later.

```

Author: > **Always store a version of the code with the time complexity to reject in the test suite**, this way, adjustments can be done easily later.

I have it in mind, I just think I will put it somewhere else (on kata authoring level, as this requirement is not specific to the test suite, but to the overall kata idea).

- **Tests should not print excessively to standard output.** Console output is limited to 1.5 MB and when the limit is exceeded, the test suite is aborted by the runner. Users sometimes debug their solutions by printing to stdout, and when the test suite itself uses up a significant fraction of the limit, users may have not enough left for debugging. This applies especially to kata which output HTML to create some visualizations: such HTML uses up the buffer space allowed for stdout output. Ideally, test suite should give detailed output **only when test fails**. Successful tests should be as quiet as possible.
- **Provide [fixed tests](#fixed-tests) as well as [random tests](#random-tests)** for the full test suite. Some test frameworks offer features like test case generators (`QuickCheck`, `ScalaCheck`), and when used, authors can choose different structure for their test suite, but still they should create test specifications in such a way it covers all necessary scenarios. See specific sections below for more details.
- **Fixed tests should cover as much edge cases as possible**. The random tests, being random aren't a placeholder to catch solutions.

> - **Random tests should _actually_ be random**, hence a random input leading to a random result. Having a batch of expecting one kind of answer than another batch expecting another behavior doesn't match what are supposed to be random tests.

Reviewer: @hobovsky Just confirming that you did not accidentally resolve this line without editing?

- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If possible, Try to compose your test suite in such a way that when test fail, they fail on some small, easy, debuggable input, before they fail on a large one, which is difficult to print. 

## Performance Tests

Some kata require solutions to be fast, or to fall into some performance category. For example, author might want to accept only solutions of complexity `O(n)`, and reject `O(n^2)` ones. Building such test suite is not always easy!

- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.

> - If difficulty of a tested kata is related to the size of the input, it's usually **better to have a few tests with large input, that thousands of tests with small input.** 200 tests with very large numbers or arrays is usually better than 10_000 tests with small ones (but see remarks on size of inputs above).

Reviewer: ```suggestion
- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with small inputs.** For example, 200 tests with huge numbers or arrays is usually better than 10,000 tests with small ones (but see remarks on size of inputs above).
```

> - **Difference beetwen accepted and rejected solutions should be easy to spot.** Ideally, tests for accepted solutions should complete very fast, and tests for rejected solutions should take very long. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful! 

Reviewer: ```suggestion
- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!
```

- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.

> ## Tests of Additional Restrictions

Reviewer: ```suggestion
## Tests with Additional Restrictions
```

> Kata may have some additional restrictions, not related to the input and output of the tested function. For example, code golf kata usually call for solutions shorter than X characters, or some kata require the task to be realized without using some class Y. Every such requirement has to be tested explicitly, but authors should be aware thet it's not always easy (or even possible):

Reviewer: ```suggestion
Some kata may impose additional restrictions on submitted solutions unrelated to its extensional behavior. For example, code golf kata usually require submitted solutions to be shorter than X characters, while others may require the task to be realized without a certain library or language feature. Every such requirement must be tested explicitly, but authors should be aware that it is difficult (or even infeasible) to do so _in a proper manner_:
```

> - If user solution has to be inspected as text, it can be found in file `/home/codewarrior/solution.txt`.

Reviewer: ```suggestion
- If the user solution should be inspected as source code, it can be found in `/home/codewarrior/solution.txt`.
```

> - Restrictions in form of _"Do not use X"_ or _"You have to use Y"_  (for example _"do not use BigInteger"_, or _"only recursive solutions are allowed"_) should not be used, because they are very difficult (sometimes even impossible) to be enforced reliably. See [this article(TODO: write the article)]() for details.

Reviewer: ```suggestion
- Restrictions of the form "do not use X" or "you must use Y" are seldom a good idea since they are extremely difficult (or downright infeasible) to reliably enforce. See [this article (TODO: write the article)]() for details.
```

## General Guidelines

- Tests are code, and as such, they should **conform to [General Coding Guidelines](/recipes/authoring/kata-snippets/coding-general/)**.

> - **Be familiar with testing framework used by Codewards for your language.** Some languages use traditional, widely known testing frameworks, like `JUnit` for Java or `chai` for JavaScript. Some use frameworks dedicated specifically for use by Codewars (for example Python). Know techniques provided by the framework (test cases generators, parametrized test cases, etc.) and assertion libraries. You can visit [reference page for your language](/languages/) to find more detailed information. 

Reviewer: ```suggestion
- **Be familiar with the testing framework used by Codewars for your language.** Some languages use traditional, widely known testing frameworks, like `JUnit` for Java or `chai` for JavaScript. Some use frameworks dedicated specifically for use by Codewars (for example Python). Know techniques provided by the framework (test case generators, parametrized test cases, etc.) and assertion libraries. You can visit the [referenced page for your language](/languages/) to find more detailed information.
```

> - **Keep amount of entries in tests report reasonably small.** Some test frameworks produce one report entry (green or red line) per test case, while some emit one entry per successful/failed assertion. Test output panel performs some DOM manipulation on the entries in client browser, and when there's a lot of them, browsers may experience performance problems. A hundred lines is usually OK, a thousand should be bearable, but a couple of thousands would be probably too much. If you need to perform many assertions and they produce more green lines a browser can handle, you can collect a result of a couple of conditions in one boolean variable and check value of this variable in a dedicated assertion.

Reviewer: ```suggestion
- **Keep the amount of entries in tests reports reasonably small.** Some test frameworks produce one report entry (green or red line) per test case, while some emit one entry per successful/failed assertion. The test output panel performs some DOM manipulation on the entries in the client's browser, and when there's a lot of them, browsers may experience performance problems. A hundred lines is usually OK, a thousand should be bearable, but a couple of thousands would probably be too much. If you need to perform many assertions and they produce more green lines a browser can handle, you can collect a result of a couple of conditions in one boolean variable and check the value of this variable in a dedicated assertion.
```

> - **Make assertion messages clear.** Some testing frameworks are notorious in creating extremely unhelpful assertion messages. `Test failed: True should equal False`  is one example of such message, and when user encounters it, it's difficult for them to tell why the test failed. Check what messages are produced by your tests (especially failed ones) and make sure they are clear for users attempting the kata, and use assertions which allow for custom messages. Something like `27 is not prime, but your solution returned true` is much better.

Reviewer: ```suggestion
- **Make assertion messages clear.** Some testing frameworks are notorious for creating extremely unhelpful assertion messages. `Test failed: True should equal False`  is one example of such a message, and when the user encounters it, it's difficult for them to tell why the test failed. Check what messages are produced by your tests (especially failed ones) and make sure they are clear for users attempting the kata, and use assertions which allow for custom messages. Something like `27 is not prime, but your solution returned true` is much better.
```

> - **Tests should not print excessively to standard output.** Console output is limited to 1.5 MB and when the limit is exceeded, test suite is aborted by the runner. Users sometimes debug their solutions by printing to stdout, and when test suite itself uses up some of the limit, users might have not enough left for debugging. This applies especially to kata which output HTML to create some visualizations: such HTML uses up the buffer space allowed for stdout output.

Reviewer: ```suggestion
- **Tests should not print excessively to standard output.** Console output is limited to 1.5 MB and when the limit is exceeded, the test suite is aborted by the runner. Users sometimes debug their solutions by printing to stdout, and when the test suite itself uses up a significant fraction of the limit, users may have not enough left for debugging. This applies especially to kata which output HTML to create some visualizations: such HTML uses up the buffer space allowed for stdout output.
```

> - Usually full test suite should contain **fixed tests**, and **random tests**. For more details, see paragraphs below.

Reviewer: ```suggestion
- Usually the full test suite should contain **fixed tests** and **random tests**. For more details, see the paragraphs below.
```

  - Input mutation by the user solution affecting the input passed to the reference solution
  - Incorrect implementation of the reference solution leading to rejection of valid user solutions
- **The reference solution should not be revealed to the user.** When an assertion fails or the test suite crashes, some testing frameworks print fragments of source code which caused the failure to the console. It may happen that such printed failure messages or stack traces expose information about the solution which should not be revealed.

> - **The reference solution shouldn't be accessible to the user solution.** It should not be possible to call reference solution directly, or implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function. Chek the [reference page for your language](/languages/) to see how to prevent this problem in your tests.

Reviewer: ```suggestion
- **The reference solution shouldn't be accessible to the user solution.** It should not be possible to call the reference solution directly, or implement the user solution as an alias or wrapper around the reference solution. The reference solution should be completely inaccessible outside the submit tests. In particular, it should not be a global and/or public function. Check the [reference page for your language](/languages/) to see how to prevent this problem in your tests.
```

Random tests are uncommon in "real life" coding and are somewhat specific to Codewars. They are required by many kata to reject solutions based on input probing, hard-coding, and other workarounds. The goal of random tests is to make the inputs unpredictable so that only proper solutions may pass.

> - **Keep amount of random tests reasonable.** If random tests are not used to verify performance, keep their amount small enough to guarantee sufficient coverage. In some cases it's not necessary to have dozens, or even hundreds of random tests, and several of them would be enough. Do not create random tests just for the sake of their amount. However, if you believe that large amount of random tests helps you to detect invalid solutions which are not caught by fixed tests, you should try and find a pattern in what types of cases fail particularly often and extract them to corresponding fixed tests.

Reviewer: ```suggestion
- **Keep the amount of random tests reasonable.** If random tests are not used to verify performance, keep their amount small enough to guarantee sufficient coverage. In some cases it's not necessary to have dozens, or even hundreds of random tests, and several of them would be enough. Do not create random tests just for the sake of their amount. However, if you believe that a large amount of random tests helps you to detect invalid solutions which are not caught by fixed tests, you should try and find a pattern in what types of cases fail particularly often and extract them to the corresponding fixed tests.
```

- **Use appropriate random utilities available in your language**. Know how to use RNGs, how to randomly generate types of inputs you need, be it numbers in a range, large numbers, strings, big integers, floating point values, collections.  Know how to do random sampling of collections, how to shuffle them, how to avoid duplicates. See this [obligatory XKCD comic](https://xkcd.com/221/) for how NOT to do random tests.
- **Random tests shouldn't be grouped by output type or behavior**, especially if the expected output is a boolean variable (e.g. checking that some input satisfies some criteria), or when it comes to error checking (solution throwing an exception in some specific situations). Hence, the order of tested scenarios should be unpredictable too. One possible way to achieve this is to generate and collect a set of random inputs for all required scenarios and shuffle them before the actual testing.

> - Under some rare circumstances, it's allowed to use so called [**randomized tests**](/glossary/#randomized-tests) instead of fully random ones. For some types of problems (for example, kata simulating chess, or problems with small set of possible inputs) it may be too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too). Before the tests are run, set of inputs should be shuffled, and eventually randomly sampled at each run, to make the hard-coding of the results more tedious for the user. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.

Reviewer: ```suggestion
- Under some rare circumstances, it's allowed to use so called [**randomized tests**](/glossary/#randomized-tests) instead of fully random ones. For some types of problems (for example, kata simulating chess, or problems with small set of possible inputs) it may be too complex or infeasible to generate inputs randomly. In such rare cases, it's accepted to use a predefined, hard-coded set of inputs (and possibly expected outputs too). Before the tests are run, the set of inputs should be shuffled, and eventually randomly sampled at each run to make the hard-coding of the results more tedious for the user. If possible, some additional transformation can be applied randomly (for example array can be reversed, game board can be rotated, sides can be flipped) if it can be easily accounted for in the result of the test.
```

- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with small inputs.** For example, 200 tests with huge numbers or arrays is usually better than 10,000 tests with small ones (but see remarks on size of inputs above).
- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!
- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.

> - **The reference solution, if used, does not have to be the same as the one in "Reference Solution" snippet.** While "Referenced Solution" snippet serves its specific purpose and is [controlled by its own set of quality guidelines](/recipes/authoring/kata-snippets/reference-solution/), reference solution used by performance tests can use different, more efficient approach, to make sure that it does not consume too much of a time limit available for user solution.

Reviewer: ```suggestion
- **The reference solution, if used, does not have to be the same as the one in the "Reference Solution" snippet.** While the "Reference Solution" snippet serves its specific purpose and is [controlled by its own set of quality guidelines](/recipes/authoring/kata-snippets/reference-solution/), the reference solution used by performance tests can use a different, more efficient approach, to make sure that it does not consume too much of a time limit available for the user solution.
```
